contributors/design-proposals/auth/pod-restriction.md

```diff
@@ -0,0 +1,477 @@
+policies that can become unmanageable.
+
+**Proposal**: Why not both? The default behavior is intersection, but you can punch a hole through
+for a specific namespace with a union approach. An optional `BindingMode` is added to the Binding
+which specifies the behavior. A pod must be valid under either:
+
+-   ANY `BindingModeAccept` policy that is bound to it, or
+
+-   ALL `BindingModeDrop` policies that are bound to it
+
+In other words, the Pod is first checked against Accept restrictions, and is accepted if any
+match. Then, it is checked against the Drop restrictions, and rejected if any don't match. The
+default behavior at the end is to accept the pod.
+
+```
+Binding:
+    Mode BindingMode
+
+BindingModeAccept
+BindingModeDrop
+```
+
+### Admission Handling
+
+The PodRestriction controller handles all create, update, and patch requests for Pod resources. It
+is also applied to all (native) resources that create pods via a PodTemplate, including ReplicaSets,
+ReplicationControllers, DaemonSets, StatefulSets, Deployments, Jobs, etc.
+
+#### Justification
+
+From a security perspective, it should be sufficient to apply the restriction to Pod requests, so
+why validate other objects? Fail-fast. By surfacing an error before the controller object is
+created, the user gets immediate feedback that they are attempting a forbidden operation. The
+alternative experience is that the controller object creation succeeds, but then the user must
+follow debug steps to determine why Pods are not being created.
+
+#### Optional Extensions
+
+New CRDs that hold a PodTemplate and create Pods should be able to be registered against the
+PodRestriction controller. The registration should include a path (auto-discovery?) to extract the
+PodTemplate from the CRD.
+
+### API Design
+
+#### Monolith vs. Modular
+
+Our current approach to policies has been "modular", i.e. a different controller and API for each
+"area" of features. This has led to some inconsistencies and gaps in our policies, and makes adding
+new restrictions needlessly complex. I propose we take a "monolithic" approach, with a single policy
+(PodRestriction) for all Pod fields.
+
+A monolithic restriction policy (i.e. single resource type) also allows the user to express more
+complex inter-relationships, such as:
+
+-   Pods using a certain volume type must run with reduced privileges
+
+-   Pods can only run as privileged if they run with a specific AntiAffinity
+
+-   Pods labeled as staging must not run on production nodes.
+
+#### No Defaulting
+
+There are 2 problems with mutating Pod fields: 1) it becomes hard to compose policies (what happens
+when the defaults conflict?) and 2) you need a policy ordering mechanism for when a Pod matches
+multiple policies. Instead, PodRestriction opts out of mutating defaults entirely, and leaves that
+for a different component to solve ([PodPreset?](../service-catalog/pod-preset.md)).
```

Comment:
What if mutation is needed via webhooks, won't they be allowed as well? If yes, it may be too restrictive.

Reply:
ok, thanks for the clarification. After reading this first line the paragraph I got confused.